### PR TITLE
feat(resource): support s3

### PR DESCRIPTION
**Pull request by saitofun**
iss #475

### CODE REVIEW

**Comment by raullenchai** [comment]
@saitofun pls add a github/clickup issue to the PR description.

**Review by saitofun — COMMENTED** [review]
TODOs:
do not read resource during `NewInstance...`, pass wasm content directly.
do not read resource repeatedly, when create applet use file data in memory when upload.

**Comment by saitofun** [comment]
> @saitofun pls add a github/clickup issue to the PR description.

added

File: pkg/depends/conf/filesystem/local/local.go [hunk 1]
```diff
@@ -33,15 +66,21 @@ func (l *LocalFileSystem) Upload(key string, data []byte) error {
...
-	return os.Remove(key)
+func (l *LocalFileSystem) Delete(md5 string) error {
+	return os.Remove(filepath.Join(l.Root, md5))
 }
 
-func isDirExists(path string) bool {
-	info, err := os.Stat(path)
-	return (err == nil || os.IsNotExist(err)) && (info != nil && info.IsDir())
+func isPathExists(path string) bool {
+	_, err := os.Stat(path)
```
Comment (huangzhiran) [inline pkg/depends/conf/filesystem/local/local.go:78]: return err == nil ?

File: pkg/depends/conf/filesystem/local/local.go [hunk 2]
```diff
@@ -33,15 +66,21 @@ func (l *LocalFileSystem) Upload(key string, data []byte) error {
 	return nil
 }
 
-func (l *LocalFileSystem) Read(key string) ([]byte, error) {
-	return os.ReadFile(key)
+func (l *LocalFileSystem) Read(md5 string) ([]byte, error) {
+	return os.ReadFile(filepath.Join(l.Root, md5))
```
Comment (huangzhiran) [inline pkg/depends/conf/filesystem/local/local.go:70]: add a new func
func (l *LocalFileSystem) RealPath(fileName string) {
    return filepath.Join(l.Root, fileName)
}

File: pkg/modules/resource/resource.go [hunk 3]
```diff
@@ -2,198 +2,207 @@ package resource
...
-func FetchOrCreateResource(ctx context.Context, accountID types.SFID, fileName string, f *multipart.FileHeader) (*models.Resource, error) {
-	d := types.MustMgrDBExecutorFromContext(ctx)
-	l := types.MustLoggerFromContext(ctx)
-	idg := confid.MustSFIDGeneratorFromContext(ctx)
-	fileSystemOp := types.MustFileSystemOpFromContext(ctx)
-
-	_, l = l.Start(ctx, "FetchOrCreateResource")
-	defer l.End()
+func Create(ctx context.Context, acc types.SFID, fh *multipart.FileHeader, md5 string) (*models.Resource, []byte, error) {
+	f, err := fh.Open()
```
Comment (huangzhiran) [inline pkg/modules/resource/resource.go:17]: need defer f.close() here ?

File: pkg/depends/conf/filesystem/local/local.go (hunk 2 above)
Comment (saitofun) [reply pkg/depends/conf/filesystem/local/local.go:70]: filename is file's md5(sum string)

**Review by huangzhiran — APPROVED** [review]

**Pull request merged**